src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -41,10 +44,59 @@ public async Task ProcessStreamAsync(Stream data, string templateDefinition, ITe
 
         public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string templateDefinition, ITelemetryLogger log)
         {
-            var template = BuildTemplate(templateDefinition, log);
-            var measurementGroups = ParseEventData(events, log);
+            // Step 1: Before processing events, validate the template
+            ILookupTemplate<IFhirTemplate> template = null;
 
-            await ProcessMeasurementGroups(measurementGroups, template, log).ConfigureAwait(false);
+            try
+            {
+                template = BuildTemplate(templateDefinition, log);
+            }
+            catch (Exception ex)
+            {
+                ex.AddEventContext(events);
+
+                if (!_exceptionTelemetryProcessor.HandleException(ex, log))
+                {
+                    throw;
+                }
+                else
+                {
+                    return; // don't continue to next step since the template is not valid
+                }
+            }
+
+            // Step 2: Transform the events into Measurements and group into Measurement Groups
+            Tuple<IEnumerable<IMeasurementGroup>, Dictionary<IMeasurement, IEventMessage>> parseEventData;
+            IEnumerable<IMeasurementGroup> measurementGroups = null;
+            Dictionary<IMeasurement, IEventMessage> measurementToEventMapping = null;
+
+            try
+            {
+                parseEventData = ParseEventData(events, log);
+                measurementGroups = parseEventData.Item1;

```

Comment:
Please use named tuples for more readable code, i.e. you can do something like ```Tuple<IEnumerable<IMeasurementGroup> groups, Dictionary<IMeasurement, IEventMessage> lookup> parseEventData```

and then do
```
measurementGroups = parseEventData.groups;
measurementToEventMapping = parseEventData.lookup;
```

Reply:
updated